src/SDiagonal.jl

```diff
@@ -0,0 +1,117 @@
+@inline (::Type{SD})(a::Tuple) where {SD <: SDiagonal} = SDiagonal(diagtype(SD)(a)) 
+@inline (::Type{SDiagonal}){N,T}(a::SVector{N,T}) = SDiagonal{N,T}(a) 
+
+@generated function SDiagonal{N,T}(a::SMatrix{N,N,T})
+    expr = [:(a[$i,$i]) for i=1:N]
+    :(SDiagonal{N,T}($(expr...)))
+end
+
+
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal{N,T}) = D
+convert{N,T}(::Type{SDiagonal{N,T}}, D::SDiagonal) = SDiagonal{N,T}(convert(SVector{N,T}, D.diag))
+
+size{N}(D::SDiagonal{N}) = (N,N)
+
+function size{N}(D::SDiagonal{N},d::Int)
+    if d<1
+        throw(ArgumentError("dimension must be ≥ 1, got $d"))
+    end
+    return d<=2 ? N : 1
+end
+
+Base.@propagate_inbounds function getindex{N,T}(D::SDiagonal{N,T}, i::Int, j::Int)  
+    @boundscheck checkbounds(D, i, j)
+    if i == j
+        @inbounds return D.diag[i]
+    else
+        zero(T)
+    end
+end
+
+# avoid linear indexing?
+Base.@propagate_inbounds function getindex{N,T}(D::SDiagonal{N,T}, k::Int) 
+    i, j = ind2sub(size(D), k)
+    D[i,j]
+end
+
+ishermitian{T<:Real}(D::SDiagonal{T}) = true
+ishermitian(D::SDiagonal) = all(D.diag .== real(D.diag))
+issymmetric(D::SDiagonal) = true
+isposdef(D::SDiagonal) = all(D.diag .> 0)
+
+factorize(D::SDiagonal) = D
+
+==(Da::SDiagonal, Db::SDiagonal) = Da.diag == Db.diag
+-(A::SDiagonal) = SDiagonal(-A.diag)
++(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag + Db.diag)
+-(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag - Db.diag)
+-(A::SDiagonal, B::SMatrix) = eye(typeof(B))*A - B
+
+*{T<:Number}(x::T, D::SDiagonal) = SDiagonal(x * D.diag)
+*{T<:Number}(D::SDiagonal, x::T) = SDiagonal(D.diag * x)
+/{T<:Number}(D::SDiagonal, x::T) = SDiagonal(D.diag / x)
+*(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag .* Db.diag)
+*(D::SDiagonal, V::SVector) = D.diag .* V
+*(V::SVector, D::SDiagonal) = D.diag .* V
+*(A::SMatrix, D::SDiagonal) = scalem(A,D.diag)
+*(D::SDiagonal, A::SMatrix) = scalem(D.diag,A)
+\(D::SDiagonal, b::SVector) = D.diag .\ b
+
+conj(D::SDiagonal) = SDiagonal(conj(D.diag))
+transpose(D::SDiagonal) = D
+ctranspose(D::SDiagonal) = conj(D)
+
+diag(D::SDiagonal) = D.diag
+trace(D::SDiagonal) = sum(D.diag)
+det(D::SDiagonal) = prod(D.diag)
+logdet{N,T<:Real}(D::SDiagonal{N,T}) = sum(log.(D.diag))
+function logdet{N,T<:Complex}(D::SDiagonal{N,T}) #Make sure branch cut is correct
+    x = sum(log.(D.diag))
+    -pi<imag(x)<pi ? x : real(x)+(mod2pi(imag(x)+pi)-pi)*im
+end
+
+eye{N,T}(::Type{SDiagonal{N,T}}) = SDiagonal(ones(SVector{N,T}))
+
+expm(D::SDiagonal) = SDiagonal(exp.(D.diag))
+logm(D::SDiagonal) = SDiagonal(log.(D.diag))
+sqrtm(D::SDiagonal) = SDiagonal(sqrt.(D.diag))
+
+\(D::SDiagonal, B::SMatrix) = scalem(1 ./ D.diag, B)
+/(B::SMatrix, D::SDiagonal) = scalem(1 ./ D.diag, B)
+\(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Db.diag ./ Da.diag)
+/(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag ./ Db.diag )
+
+function inv{N,T}(D::SDiagonal{N,T})
+    for i = 1:N
```

Comment:
Also, we generally unroll such loops elsewhere... 

It's also slightly annoying that the code to check if we should throw a singular exception will be slower than taking the inverse itself... but I guess `Base.Diagonal` does this too?